### PR TITLE
Login Backend

### PR DESCRIPTION
**Pull request by abedshamia**
Validate the incoming request.
Check if the user exists in collection
Compare Password
Check if user is verified
Send back a cookie with a token
Tested login cases

### CODE REVIEW

File: server/controllers/authController.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+  await loginValidation(req.body);
+
+  const user = await User.findOne({ email });
+
+  if (!user) {
+    throw new CustomError('User not found', 404);
+  }
+
+  const validPassword: boolean = await bcrypt.compare(password, user.password);
```
Comment (Raghad-Mezied) [inline server/controllers/authController.ts:21]: Destructure the compare.

File: server/controllers/authController.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+  if (!validPassword) {
+    throw new CustomError('Invalid password', 400);
+  }
+
+  if (!user.is_confirmed) {
+    throw new CustomError('Please Verify your email', 401);
+  }
+
+  const token: string = sign({ id: user.id }, process.env.JWT_SECRET ?? '', { expiresIn: '1h' });
```
Comment (Raghad-Mezied) [inline server/controllers/authController.ts:31]: You are using sign sync, so convert it to promise in the utils and use it here.
And restructure all your variables (JWT_SECRET, id, and is_confirmed)

File: server/controllers/authController.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+  res.cookie('jwt', token, {
```
Comment (Raghad-Mezied) [inline server/controllers/authController.ts:33]: Name it token instead of jwt

File: server/controllers/authController.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    message: 'Login successful',
+    data: {
+      user: {
+        id: user.id,
+        email: user.email,
+        role: user.role,
+      },
+      token,
+    },
+  });
```
Comment (Raghad-Mezied) [inline server/controllers/authController.ts:48]: Destructure the id, email, and the role. And why do you need to send the token?

File: server/controllers/authController.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
+/* eslint-disable import/prefer-default-export */
+/* eslint-disable no-unused-vars */
```
Comment (Raghad-Mezied) [inline server/controllers/authController.ts:2]: Don't disable `eslint`, solve them

File: server/controllers/authController.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import { Request, Response } from 'express';
+import { sign } from 'jsonwebtoken';
+import bcrypt from 'bcrypt';
+import loginValidation from '../utils/validation/loginValidation';
```
Comment (Raghad-Mezied) [inline server/controllers/authController.ts:6]: Add the validation in the index and here import it with the CustomError from the index

File: server/controllers/authController.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  }
+
+  const token: string = sign({ id: user.id }, process.env.JWT_SECRET ?? '', { expiresIn: '1h' });
+
+  res.cookie('jwt', token, {
+    httpOnly: true,
+    maxAge: 1000 * 60 * 60,
+  });
+
+  return res.status(200).json({
```
Comment (Raghad-Mezied) [inline server/controllers/authController.ts:38]: As 200 is the default we don't need to send it.

File: server/database/Models/User.ts [hunk 8]
```diff
@@ -37,6 +37,11 @@ const user = new Schema({
     required: true,
     enum: ['admin', 'interviewer', 'interviewee'],
   },
+  is_confirmed: {
```
Comment (Raghad-Mezied) [inline server/database/Models/User.ts:40]: Just name it is_verified as Mahmoud did

File: server/routes/authRoutes.ts [hunk 9]
```diff
@@ -0,0 +1,8 @@
+import express from 'express';
+import { login } from '../controllers';
+
+const router: any = express.Router();
```
Comment (Raghad-Mezied) [inline server/routes/authRoutes.ts:4]: Edit the any type

File: server/utils/validation/loginValidation.ts [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+  const schema = Joi.object({
+    email: Joi.string().email().required(),
+    password: Joi.string().required().regex(/^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{4,}$/).message('Password must contain at least 4 characters, one upper case letter, one lower case letter, one number, and one special character'),
+    role: Joi.string().required(),
+  });
+
+  const result: any = await schema.validateAsync(data);
+  return result;
+};
+
```
Comment (Raghad-Mezied) [inline server/utils/validation/loginValidation.ts:13]: Edit any type

File: server/tests/routes.test.ts [hunk 11]
```diff
@@ -1,6 +1,89 @@
...
+      password: 'Abed@123',
+      role: 'interviewee',
+    }).expect(200)
+      .end((err, res) => {
+        if (err) {
+          return done(err);
+        }
+        expect(res.body.message).toBe('Login successful');
+        expect(res.body.data.user.email).toBe('jack@gmail.com');
+        return done();
```
Comment (Raghad-Mezied) [inline server/tests/routes.test.ts:84]: Test if the route set a cookie

**Review by Raghad-Mezied — CHANGES_REQUESTED** [review]
Good Job Abedalrahman

File: server/controllers/authController.ts [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    throw new CustomError('Please Verify your email', 401);
+  }
+
+  const payload: JwtPayload = {
+    id,
+    email,
+    role,
+  };
+
+  const token: any = await signToken(payload, JWT_SECRET, { expiresIn: '1h' });
```
Comment (Raghad-Mezied) [inline server/controllers/authController.ts:44]: As we can access the JWT_SECRET in the utils, then no need to pass it here just use it directly in your function.
And edit the `any` type.

File: server/controllers/authController.ts [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+      user: {
+        id,
+        email,
+        role,
+      },
+    },
+  });
+};
+
+export default login;
```
Comment (Raghad-Mezied) [inline server/controllers/authController.ts:63]: As we will need other auth controllers so add a folder for them call it auth and inside it, you can put your file.

File: server/utils/index.ts [hunk 14]
```diff
@@ -1,4 +1,6 @@
 /* eslint-disable import/prefer-default-export */
```
Comment (Raghad-Mezied) [inline server/utils/index.ts:1]: You already solve this problem so delete this line

File: server/database/build.ts [hunk 15]
```diff
@@ -1,47 +1,48 @@
...
-    console.log('Fake data created successfully');
-  } catch (err: any) {
-    throw new Error(err);
-  }
-};
-
-if (process.env.NODE_ENV === 'development') {
-  createFakeData();
-}
+/* eslint-disable linebreak-style */
```
Comment (Raghad-Mezied) [inline server/database/build.ts:1]: Don't disable `eslint`, and you can use `lint-fix` script to fix all these styles.

**Review by Raghad-Mezied — CHANGES_REQUESTED** [review]

**Review by manar-abed — APPROVED** [review]

**Review by Raghad-Mezied — APPROVED** [review]

**Review by Mahmoud-Ahmad2 — APPROVED** [review]

**Pull request merged**